### PR TITLE
test(clp-s): Add end-to-end test case for compression and extraction.

### PR DESCRIPTION
**Pull request by AVMatthews**
# Description

Adds a test case that runs the full clp-s pipeline from JSON to archive and back to JSON to ensure data consistency. Adds corresponding example JSON files to test on. 

Notes: 
- Due to discrepancies in the number of significant digits produces in the reconstruction of the JSON file, the current JSON file used in the test, does not contain floats. I leave testing of float value accuracy to a later time.
-  The values of int64_max_jq_losslessly_represents and int64_min_jq_losslessly_represents in json file used for testing is not utilizing the full 2^63 value bits of a int64 because jq does not support full 64 bit integers. Instead it supports integers in the range of -2^53 ... 2^53. 125276004817190914. Integers out of this range it approximates (https://github.com/jqlang/jq/issues/369).

# Validation performed

Compiled and ran the test to ensure that the REQUIRE statements are met and manually checked output files to see that the correct data is produced.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new end-to-end testing framework for the `clp_s` module, focusing on JSON data compression and extraction.
	- Added numerous source files to enhance functionality within the `clp_s` module.

- **Bug Fixes**
	- Enhanced package installation scripts for CentOS and Ubuntu to include necessary tools (`diffutils` and `jq`).

- **Tests**
	- Added a comprehensive test file to validate the `clp_s` module's performance and accuracy.

- **Chores**
	- Updated package installation scripts across multiple environments to ensure consistency and availability of required tools.

### CODE REVIEW

File: components/core/tests/test-end_to_end.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        four_byte_encoded_variable_t,
+        eight_byte_encoded_variable_t
+) {
+    std::filesystem::remove_all("test-end-to-end-archive");
+    std::filesystem::remove_all("test-end-to-end-out");
+    std::filesystem::remove("test-end-to-end_sorted.json");
+    std::filesystem::remove("diff_out.txt");
+
+    std::filesystem::create_directory("test-end-to-end-archive");
+    REQUIRE(std::filesystem::is_directory("test-end-to-end-archive"));
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test-end_to_end.cpp:65]: I am not trying to remove these files at the end for clean up, I"m trying to remove them if present before I run the test-case, so any current content of the files/directory don't influence the test

File: components/core/tests/test_log_files/test_sorted.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+{"clp_string":"uid=0, CPU usage:99.99%, \"user_name\"=YScope","empty_object":{},"false":false,"float_neg":-1.01,"float_pos":1.01,"float_zero":0.0,"int16_max":32767,"int16_min":-32768,"int32_max":2147483647,"int32_min":-2147483648,"int64_max_supported":9223372036854776,"int64_min_supported":-9223372036854776,"int8_max":127,"int8_min":-128,"nonempty_array":[1,2,3,4,5],"null":null,"string":"short_string","true":true}
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test_log_files/test_sorted.json:4]: I'll consider adding more edge cases at another time.

File: components/core/tests/test-end_to_end.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    parser_option.print_archive_stats = cDefaultPrintArchiveStats;
+    parser_option.structurize_arrays = cDefaultStructurizeArrays;
+
+    clp_s::JsonParser parser(parser_option);
+    REQUIRE(parser.parse());
+    parser.store();
+
+    REQUIRE(false == std::filesystem::is_empty("test-end-to-end-archive"));
+
+    std::filesystem::create_directory("test-end-to-end-out");
```
Comment (gibber9809) [inline components/core/tests/test-end_to_end.cpp:83]: All of these hardcoded paths and directories should probably be constants and live near the other constants at the top of this file.

`constexpr char cTestEndToEndOutputDirectory[] = "test-end-to-end-out";` etc.

File: components/core/tests/test-end_to_end.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/tests/test-end_to_end.cpp:65]: I'm inclined to agree with rabbit here. You don't need to to something that subclasses Catch::TestEventListener, but you could do something simple that cleans things up before/after every test.

```
namespace {
class EndToEndTestFixture {
    EndToEndTestFixture() { // setup }
    ~EndToEndTestFixture() { // teardown }
};
} // namespace
```

And then all you need to do is put `EndToEndTestFixture const fixture{};` At the top of your TEST_CASE (but after the GENERATE stuff I mentioned).

File: components/core/tests/test-end_to_end.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    REQUIRE(false == std::filesystem::is_empty("test-end-to-end_sorted.json"));
+
+    result = std::system("command -v diff >/dev/null 2>&1");
+    REQUIRE(0 == result);
+    std::string const command = "diff -u test-end-to-end_sorted.json " + get_test_input_local_path()
+                                + " > diff_out.txt";
+    result = std::system(command.c_str());
+    REQUIRE((0 == WEXITSTATUS(result) || 1 == WEXITSTATUS(result)));
+
+    REQUIRE(std::filesystem::is_empty("diff_out.txt"));
```
Comment (gibber9809) [inline components/core/tests/test-end_to_end.cpp:120]: I don't like producing side effects from tests, though I do see why you want to leave this diff file around. Even still, we should at least make sure this file gets deleted when the test is successful.

File: components/core/tests/test-end_to_end.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    parser.store();
+
+    REQUIRE(false == std::filesystem::is_empty("test-end-to-end-archive"));
+
+    std::filesystem::create_directory("test-end-to-end-out");
+    REQUIRE(std::filesystem::is_directory("test-end-to-end-out"));
+
+    clp_s::JsonConstructorOption constructor_option{};
+    constructor_option.output_dir = "test-end-to-end-out";
+    constructor_option.ordered = false;
```
Comment (gibber9809) [inline components/core/tests/test-end_to_end.cpp:88]: Now that ordered compression/decompression is merged is it worth considering using ordered compression/decompression for these tests?

File: components/core/tests/test-end_to_end.cpp (hunk 4 above)
Comment (gibber9809) [reply components/core/tests/test-end_to_end.cpp:120]: Actually on second thought it's probably best to just make sure everything gets deleted via fixture (including this diff) regardless of success or failure just to keep things simple. It should be relatively easy to reproduce failures by just looking at the test and running the commands.

**Review by gibber9809 — CHANGES_REQUESTED** [review]

File: components/core/tests/test-end_to_end.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+#include <Catch2/single_include/catch2/catch.hpp>
+#include <msgpack.hpp>
+
+#include "../src/clp/BufferReader.hpp"
+#include "../src/clp/ffi/ir_stream/decoding_methods.hpp"
+#include "../src/clp/ffi/ir_stream/Deserializer.hpp"
+#include "../src/clp/ffi/ir_stream/Serializer.hpp"
+#include "../src/clp/ffi/KeyValuePairLogEvent.hpp"
+#include "../src/clp/ir/types.hpp"
+#include "../src/clp/time_types.hpp"
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test-end_to_end.cpp:16]: Sorry I forgot to remove these. I have another test-case which isn't ready that I stripped out before this PR and I forgot to double check the includes.

File: components/core/tools/scripts/lib_install/centos-stream-9/install-prebuilt-packages.sh [hunk 7]
```diff
@@ -8,9 +8,11 @@ set -u
 
 dnf install -y \
     cmake \
+    diffutils \
     gcc-c++ \
     git \
     java-11-openjdk \
+    jq \
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tools/scripts/lib_install/centos-stream-9/install-prebuilt-packages.sh:15]: It seemed to be installed by default on macOS where it's not on ubuntu and  centos.  I don't think there is even a brew formula for regular diff in macOS.

**Review by gibber9809 — DISMISSED** [review]
LGTM. I think one thing we may want to do is rename `test-end_to_end.cpp` -> `test-clp_s_end_to_end.cpp` or something similar, but I'll leave that choice up to you.

Also not sure if @kirkrodrigues or anyone else wants to take a look before this gets merged?

File: components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh [hunk 8]
```diff
@@ -12,11 +12,13 @@ DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
   checkinstall \
   cmake \
   curl \
+  diffutils \
```
Comment (kirkrodrigues) [inline components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh:15]: I see `diff` is already installed by default in Ubuntu containers, so do we need this?

File: components/core/tests/test-end_to_end.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+#include <filesystem>
+#include <string>
+#include <vector>
+
+#include <Catch2/single_include/catch2/catch.hpp>
+
+#include "../src/clp_s/JsonConstructor.hpp"
+#include "../src/clp_s/JsonParser.hpp"
+
+auto const cDefaultTargetEncodedSize = 8ULL * 1024 * 1024 * 1024;  // 8 GB
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:13]: Let's use the appropriate SI units (e.g., `GiB`).

File: components/core/tests/test-end_to_end.cpp [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+auto const cDefaultCompressionLevel = 3;
+auto const cDefaultPrintArchiveStats = false;
+auto const cDefaultOrdered = false;
+auto const cDefaultTargetOrderedChunkSize = 0;
+
+std::string const cTestEndToEndArchiveDirectory{"test-end-to-end-archive"};
+std::string const cTestEndToEndOutputDirectory = "test-end-to-end-out";
+std::string const cTestEndToEndOutputSortedJson = "test-end-to-end_sorted.json";
+std::string const cTestEndToEndInputFileDirectory = "test_log_files";
+std::string const cTestEndToEndInputFile = "test_no_floats_sorted.json";
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:25]: 1. As mentioned in the [style guide](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables), we shouldn't use static init for non-trivial objects. Any reason not to use a `constexpr string_view` as mentioned in our internal guidelines?
2. We should use brace-init for all of these.
3. Since these are only used in the one test, I would just put these directly in the test that they're used. When we add more tests later, we can easily move them out to be globals.

File: components/core/tests/test-end_to_end.cpp [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace {
+
+class Cleanup {
+public:
+    Cleanup() {
+        std::filesystem::remove_all(cTestEndToEndArchiveDirectory);
+        std::filesystem::remove_all(cTestEndToEndOutputDirectory);
+        std::filesystem::remove(cTestEndToEndOutputSortedJson);
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:34]: Can we move this into a static private method that we call from both the constructor and destructor?

File: components/core/tests/test-end_to_end.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+std::string const cTestEndToEndArchiveDirectory{"test-end-to-end-archive"};
+std::string const cTestEndToEndOutputDirectory = "test-end-to-end-out";
+std::string const cTestEndToEndOutputSortedJson = "test-end-to-end_sorted.json";
+std::string const cTestEndToEndInputFileDirectory = "test_log_files";
+std::string const cTestEndToEndInputFile = "test_no_floats_sorted.json";
+
+namespace {
+
+class Cleanup {
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:29]: Need a docstring.

File: components/core/tests/test-end_to_end.cpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+auto const cDefaultTargetOrderedChunkSize = 0;
+
+std::string const cTestEndToEndArchiveDirectory{"test-end-to-end-archive"};
+std::string const cTestEndToEndOutputDirectory = "test-end-to-end-out";
+std::string const cTestEndToEndOutputSortedJson = "test-end-to-end_sorted.json";
+std::string const cTestEndToEndInputFileDirectory = "test_log_files";
+std::string const cTestEndToEndInputFile = "test_no_floats_sorted.json";
+
+namespace {
+
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:28]:
```suggestion
```

Unless the line is necessary?

File: components/core/tests/test-end_to_end.cpp [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    parser_option.file_paths.push_back(get_test_input_local_path());
+    parser_option.archives_dir = cTestEndToEndArchiveDirectory;
+    parser_option.target_encoded_size = cDefaultTargetEncodedSize;
+    parser_option.max_document_size = cDefaultMaxDocumentSize;
+    parser_option.min_table_size = cDefaultMinTableSize;
+    parser_option.compression_level = cDefaultCompressionLevel;
+    parser_option.print_archive_stats = cDefaultPrintArchiveStats;
+    parser_option.structurize_arrays = structurize_arrays;
+
+    clp_s::JsonParser parser(parser_option);
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:74]: Brace init. Double-check all your object creations.

File: components/core/tests/test-end_to_end.cpp [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        clp_s::JsonConstructor constructor(constructor_option);
+        constructor.store();
+    }
+
+    std::string command = cTestEndToEndOutputDirectory + "/original";
+    REQUIRE(std::filesystem::exists(command.c_str()));
+
+    int result = std::system("command -v jq >/dev/null 2>&1");
+    REQUIRE(0 == result);
+    command = "jq -S -c '.' " + cTestEndToEndOutputDirectory + "/original | sort > "
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:104]: 1. Let's use the long-form for all these options, for readability.
2. Let's use std::format instead of string concatenation to improve readability.

File: components/core/tests/test-end_to_end.cpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+            // Skip non-directories
+            continue;
+        }
+
+        constructor_option.archive_id = entry.path().filename();
+        clp_s::JsonConstructor constructor(constructor_option);
+        constructor.store();
+    }
+
+    std::string command = cTestEndToEndOutputDirectory + "/original";
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:99]: Can we use filesystem::path instead of manually building a path string?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Some quick comments. Will do a more thorough review after you address these comments.

File: components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh (hunk 8 above)
Comment (AVMatthews) [reply components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh:15]: That's fine by me. I had added it while debugging, just to be sure, but we can drop it now. Is it default in centos as well or should I leave that one?

Comment (kirkrodrigues) [reply components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh:15]: `diff` used to be included in the centos container but it's not anymore. You can run the container with: `docker run --rm -it dokken/centos-stream-9 /bin/bash` and then check.

File: components/core/tests/test-end_to_end.cpp (hunk 10 above)
Comment (AVMatthews) [reply components/core/tests/test-end_to_end.cpp:25]: 3. These are used in the class and the 2 helper functions. Personally, I think it's cleaner to leave it at the top, so they don't need to be declared or passed to so many places.

Comment (kirkrodrigues) [reply components/core/tests/test-end_to_end.cpp:25]: 3. Sorry, this comment was supposed to apply to the variables above.

File: components/core/tests/test-end_to_end.cpp (hunk 14 above)
Comment (kirkrodrigues) [reply components/core/tests/test-end_to_end.cpp:74]: No, no, I'm saying double check that you're using brace-init wherever you're initializing an object. Like this place is one instance.

File: components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh (hunk 8 above)
Comment (AVMatthews) [reply components/core/tools/scripts/lib_install/ubuntu-focal/install-prebuilt-packages.sh:15]: Yeah, not in centos. Did you check both versions of ubuntu? Should I take it out of both of them?

**Comment by kirkrodrigues** [comment]
Can you merge main to see if that resolves the current macOS build issue?

**Comment by kirkrodrigues** [comment]
Since macOS 13 has trouble with std::format, can you try `fmt::format`?

File: components/core/tests/test-end_to_end.cpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+auto get_test_input_path_relative_to_tests_dir() -> std::filesystem::path {
+    return std::filesystem::path{cTestEndToEndInputFileDirectory} / cTestEndToEndInputFile;
+}
+
+auto get_test_input_local_path() -> std::string {
+    std::filesystem::path const current_file_path{__FILE__};
+    auto const tests_dir{current_file_path.parent_path()};
+    return (tests_dir / get_test_input_path_relative_to_tests_dir()).string();
+}
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:48]: Although trivial, we should still follow best practices and separate function declarations from their definitions (i.e., put the prototypes for these functions above the functions so that if for instance, we reorder the functions, we don't get a compile error).

File: components/core/tests/test-end_to_end.cpp [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+}  // namespace
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("clp-s_compression_and_extraction_no_floats", "[clp-s][end-to-end]") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:52]:
```suggestion
TEST_CASE("clp-s-compress-extract-no-floats", "[clp-s][end-to-end]") {
```

I think that's more compact but with the same meaning?

File: components/core/tests/test-end_to_end.cpp [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    auto const default_target_encoded_size = 8ULL * 1024 * 1024 * 1024;  // 8 GiB
+    auto const default_max_document_size = 512ULL * 1024 * 1024;  // 512 MiB
+    auto const default_min_table_size = 1ULL * 1024 * 1024;  // 1 MiB
+    auto const default_compression_level = 3;
+    auto const default_print_archive_stats = false;
+    auto const default_ordered = false;
+    auto const default_target_ordered_chunk_size = 0;
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:59]: These should be constexpr.

Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:59]: These are unused?

File: components/core/tests/test-end_to_end.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    auto structurize_arrays = GENERATE(true, false);
+
+    Cleanup const test_cleanup;
+
+    std::filesystem::create_directory(cTestEndToEndArchiveDirectory);
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:64]: Can we break this function up into compress, extract, and compare methods? I think that would make it easier to read quickly.

File: components/core/tests/test-end_to_end.cpp [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    parser_option.min_table_size = default_min_table_size;
+    parser_option.compression_level = default_compression_level;
+    parser_option.print_archive_stats = default_print_archive_stats;
+    parser_option.structurize_arrays = structurize_arrays;
+
+    clp_s::JsonParser parser{parser_option};
+    REQUIRE(parser.parse());
+    parser.store();
+
+    REQUIRE(false == std::filesystem::is_empty(cTestEndToEndArchiveDirectory));
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:81]: To resolve the clang-tidy warning, everywhere we do a `REQUIRE(x <op> y)` operation, we should wrap the inner statement in an extra set of parentheses.

File: components/core/tests/test-end_to_end.cpp [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        constructor_option.archive_id = entry.path().filename();
+        clp_s::JsonConstructor constructor{constructor_option};
+        constructor.store();
+    }
+
+    std::filesystem::path extracted_json_path{cTestEndToEndOutputDirectory};
+    extracted_json_path /= "original";
+    REQUIRE(std::filesystem::exists(extracted_json_path));
+
+    int result = std::system("command -v jq >/dev/null 2>&1");
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:106]: For the two `command` invocations, do we need the `2>&1`?

File: components/core/tests/test-end_to_end.cpp [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    REQUIRE(0 == result);
+    std::string command = fmt::format(
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:108]:
```suggestion
    auto command = fmt::format(
```

Not absolutely necessary, but is simpler to write.

File: components/core/tests/test-end_to_end.cpp [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+            "jq --sort-keys --compact-output '.' {}/original | sort > {}",
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:109]: We can use `extracted_json_path.string()` instead of `{}/original`, right?

File: components/core/tests/test-end_to_end.cpp [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    );
+    result = std::system(command.c_str());
+    REQUIRE(0 == result);
+
+    REQUIRE(false == std::filesystem::is_empty(cTestEndToEndOutputSortedJson));
+
+    result = std::system("command -v diff >/dev/null 2>&1");
+    REQUIRE(0 == result);
+    command = fmt::format(
+            "diff -u {} {}  > /dev/null",
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:121]:
```suggestion
            "diff --unifiied {} {} > /dev/null",
```

File: components/core/tests/test-end_to_end.cpp [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+            cTestEndToEndOutputSortedJson,
+            get_test_input_local_path()
+    );
+    result = std::system(command.c_str());
+    REQUIRE(0 == WEXITSTATUS(result));
```
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:126]: The [docs](https://pubs.opengroup.org/onlinepubs/9699919799/functions/wait.html) for `WEXITSTATUS` suggest that it's only valid to use `WEXITSTATUS` if we check `WIFEXITED` first, right?

File: components/core/tests/test-end_to_end.cpp (hunk 18 above)
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:52]: Using `std::system` triggers two clang-tidy warnings: `cert-env33-c` and `concurrency-mt-unsafe`. We should silence both with something like:

```cpp
    // Silence the checks below since our use of `std::system` is safe in the context of testing.
    // NOLINTBEGIN(cert-env33-c,concurrency-mt-unsafe)
```

File: components/core/tests/test_log_files/test_no_floats_sorted.json [hunk 27]
```diff
@@ -0,0 +1,4 @@
+{"clp_string":"uid=0, CPU usage:99.99%, \"user_name\"=YScope","empty_array":[],"empty_object":{},"false":false,"int16_max":32767,"int16_min":-32768,"int32_max":2147483647,"int32_min":-2147483648,"int64_max_supported":9223372036854776,"int64_min_supported":-9223372036854776,"int8_max":127,"int8_min":-128,"null":null,"string":"short_string","true":true}
```
Comment (kirkrodrigues) [inline components/core/tests/test_log_files/test_no_floats_sorted.json:1]: Can you mention in the PR description why int64_max/min_supported is less than 2^63?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-end_to_end.cpp (hunk 23 above)
Comment (AVMatthews) [reply components/core/tests/test-end_to_end.cpp:108]: I personally like leaving the types if I know it. I dislike tracking down the type when modifying other peoples code. If you have a strong preference I will change it.

Comment (kirkrodrigues) [reply components/core/tests/test-end_to_end.cpp:108]: Sure, then I would suggest switching to the `std::string command{fmt::format(...)}` way of writing this.

That said:
* I would imagine VSCode has something similar to CLion where it can display the types for `auto` variables.
* There are advantages to using `auto` in that we don't unintentionally convert the return type to the type we think the method returns (e.g., if the method changes its signature at some time later).
* Sometimes we don't absolutely need to know the return type---e.g., in this case, we only require that the returned object has a `c_str()` method, not that it has to be a `std::string`.
* C++ with templates can be a *very* verbose language, so using `auto` can make the code more readable (while IDE type hints and inline docs can step in when we really need to know the underlying type).

File: components/core/tests/test-end_to_end.cpp (hunk 22 above)
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:106]:
```suggestion
    int result{std::system("command -v jq >/dev/null 2>&1")};
```

Forgot this in the last review.

File: components/core/tests/test-end_to_end.cpp
Comment (kirkrodrigues) [inline components/core/tests/test-end_to_end.cpp:1]: Forgot to mention this in the last review, but can we rename this to `test-clp_s-end_to_end.cpp`?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-end_to_end.cpp (hunk 22 above)
Comment (AVMatthews) [reply components/core/tests/test-end_to_end.cpp]: It should redirect both stderr and stdout to /dev/null. I'm not sure if it's strictly necessary, but I don't think it is harmful.

File: components/core/tests/test-clp_s-end_to_end.cpp [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+    parser_option.max_document_size = cDefaultMaxDocumentSize;
+    parser_option.min_table_size = cDefaultMinTableSize;
+    parser_option.compression_level = cDefaultCompressionLevel;
+    parser_option.print_archive_stats = cDefaultPrintArchiveStats;
+    parser_option.structurize_arrays = structurize_arrays;
+
+    clp_s::JsonParser parser{parser_option};
+    REQUIRE(parser.parse());
+    parser.store();
+
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test-clp_s-end_to_end.cpp:85]: We check this with the require statements

File: components/core/tests/test-clp_s-end_to_end.cpp [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+    for (auto const& entry : std::filesystem::directory_iterator(constructor_option.archives_dir)) {
+        if (false == entry.is_directory()) {
+            // Skip non-directories
+            continue;
+        }
+
+        constructor_option.archive_id = entry.path().filename();
+        clp_s::JsonConstructor constructor{constructor_option};
+        constructor.store();
+    }
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test-clp_s-end_to_end.cpp:110]: we do have some checking using the require segments and later comparison

File: components/core/tests/test-clp_s-end_to_end.cpp [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+    result = std::system("command -v diff >/dev/null 2>&1");
+    REQUIRE((0 == result));
+    command = fmt::format(
+            "diff --unified {} {}  > /dev/null",
+            cTestEndToEndOutputSortedJson,
+            get_test_input_local_path()
+    );
+    result = std::system(command.c_str());
+    REQUIRE((true == WIFEXITED(result)));
+    REQUIRE((0 == WEXITSTATUS(result)));
```
[the remark this answers is not in the document]
Comment (AVMatthews) [reply components/core/tests/test-clp_s-end_to_end.cpp:142]: According the c++ documentation this is the correct way the check std::system

File: components/core/tests/test-clp_s-end_to_end.cpp [hunk 31]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// NOLINTEND(cert-env33-c,concurrency-mt-unsafe)
+}  // namespace
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
```
Comment (kirkrodrigues) [inline components/core/tests/test-clp_s-end_to_end.cpp:148]:
```suggestion
```

No longer necessary since we broke the function up.

File: components/core/tests/test-clp_s-end_to_end.cpp [hunk 32]
```diff
@@ -0,0 +1,159 @@
...
+TEST_CASE("clp-s-compress-extract-no-floats", "[clp-s][end-to-end]") {
+    auto structurize_arrays = GENERATE(true, false);
+
+    TestOutputCleaner const test_cleanup;
+
+    compress(structurize_arrays);
+
+    std::filesystem::path extracted_json_path = extract();
```
Comment (kirkrodrigues) [inline components/core/tests/test-clp_s-end_to_end.cpp:156]:
```suggestion
    std::filesystem::path const extracted_json_path{extract()};
```

Or use `auto`.

**Review by kirkrodrigues — COMMENTED** [review]
> The values of int64_max_supported and int_64_min_supported in json file used for testing is not utilizing the full 2^63 value bits because jq does not support full 64 bit integers.

Does that imply that the values you're using are the extremes that `jq`' supports?

**Comment by AVMatthews** [comment]
> > The values of int64_max_supported and int_64_min_supported in json file used for testing is not utilizing the full 2^63 value bits because jq does not support full 64 bit integers.
> 
> Does that imply that the values you're using are the extremes that `jq`' supports?

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

```
test(clp-s): Add end-to-end test case for compression and extraction.
```
